Alex and Happy Part 1/article.md

```diff
@@ -0,0 +1,842 @@
+  , RangedToken {rtToken = Identifier "true", rtRange = Range {start = AlexPn 3 1 4, stop = AlexPn 7 1 8}}
+  , RangedToken {rtToken = Then, rtRange = Range {start = AlexPn 8 1 9, stop = AlexPn 12 1 13}}
+  , RangedToken {rtToken = Identifier "foo", rtRange = Range {start = AlexPn 13 1 14, stop = AlexPn 16 1 17}}
+  , RangedToken {rtToken = Else, rtRange = Range {start = AlexPn 17 1 18, stop = AlexPn 21 1 22}}
+  , RangedToken {rtToken = LPar, rtRange = Range {start = AlexPn 22 1 23, stop = AlexPn 23 1 24}}
+  , RangedToken {rtToken = Identifier "bar", rtRange = Range {start = AlexPn 23 1 24, stop = AlexPn 26 1 27}}
+  , RangedToken {rtToken = Identifier "baz", rtRange = Range {start = AlexPn 27 1 28, stop = AlexPn 30 1 31}}
+  , RangedToken {rtToken = RPar, rtRange = Range {start = AlexPn 30 1 31, stop = AlexPn 31 1 32}}
+  , RangedToken {rtToken = EOF, rtRange = Range {start = AlexPn 31 1 32, stop = AlexPn 31 1 32}}
+  ]
+```
+
+### Lexing integers
+
+Lexing integers is pretty straightforward. We can just use a sequence of one or more digits.
+
+```alex
+-- Identifiers
+<0> @id     { tokId }
+
+-- Constants
+<0> $digit+ { tokInteger }
+```
+
+And of course, we need to define `tokInteger`:
+
+```haskell
+tok = ...  -- anchor
+
+tokInteger :: AlexAction RangedToken
+tokInteger inp@(_, _, str, _) len =
+  pure RangedToken
+    { rtToken = Integer $ read $ BS.unpack $ BS.take len str
+    , rtRange = mkRange inp len
+    }
+```
+
+Check that it works:
+
+```haskell
+>>> scanMany "42"
+Right
+  [ RangedToken {rtToken = Integer 42, rtRange = Range {start = AlexPn 0 1 1, stop = AlexPn 2 1 3}}
+  , RangedToken {rtToken = EOF, rtRange = Range {start = AlexPn 2 1 3, stop = AlexPn 2 1 3}}
+  ]
+```
+
+#### Lexing comments
+
+Lexing comments will require us to use an extra start code besides just `0`.
+Surely, we could make something as simple as the following:
+
+```alex
+<0> $white+ ;
+
+<0> "(*" (. | \n)* "*)" ;
+
+-- Keywords
+```
+
+But this will lead to some problems.
+For instance, what if the comment is unclosed?
+
+```haskell
+>>> scanMany "(* my comment"
+Right
+  [ RangedToken {rtToken = LPar, rtRange = Range {start = AlexPn 0 1 1, stop = AlexPn 1 1 2}}
+  , RangedToken {rtToken = Times, rtRange = Range {start = AlexPn 1 1 2, stop = AlexPn 2 1 3}}
+  , RangedToken {rtToken = Identifier "my", rtRange = Range {start = AlexPn 3 1 4, stop = AlexPn 5 1 6}}
+  , RangedToken {rtToken = Identifier "comment", rtRange = Range {start = AlexPn 6 1 7, stop = AlexPn 13 1 14}}
+  , RangedToken {rtToken = EOF, rtRange = Range {start = AlexPn 13 1 14, stop = AlexPn 13 1 14}}
+  ]
+```
+
+We would like to detect such cases and also support nested comments.
+We will create two helper functions, `nestComment` and `unnestComment`, which will keep track of how many layers of comments we have and allow us to detect whether we have reached the end of the file with an unclosed comment.
```

Comment:
People might go searching for these functions, would be good to give them a little hint that you're building up to them.